### PR TITLE
MLM compatibility

### PR DESCRIPTION
**Pull request by grezniczek**
This makes this module compatible with MLM.
Due to making use of the `afterRender()` framework function, the minimal REDCap version supported is 12.2.10.

### CODE REVIEW

File: ImageMapClass.php
Comment (steentje34) [inline ImageMapClass.php]: Just a question: yesterday, I also took a shot at trying to solve the issue, and I was trying to come up with a solution that would function in REDCap versions before v12.0.0, between v12.0.0 and v12.2.10 and after v12.2.10. I came up with the solution below (requiring no changes to the .js files, and only minimal changes to the config.json), but haven't thoroughly tested yet. Do you think this could also work? If yes, do you see advantages/disadvantages linked to both solutions?

As of line 19:
```php
                if ($redcap_version >= '12.2.10') {
                    $keyLabelCodeMap[$fieldname]['script'] = "module.afterRender(function() {try{" . $jsCall . '}catch(e) {console.warn(e);}})';
                } elseif ($redcap_version >= '12.0') {
                    $keyLabelCodeMap[$fieldname]['script'] = "REDCap.MultiLanguage.onLangChanged(function() {try{" . $jsCall . '}catch(e) {console.warn(e);}})';
                } else {
                    $keyLabelCodeMap[$fieldname]['script'] = "try{" . $jsCall . '}catch(e) {console.warn(e);}';
                }
            }
        }
        
        $this->initializeJavascriptModuleObject();
        echo '<script>';
        print file_get_contents(__DIR__ . "/imagemapfunctions.js");
        print file_get_contents(__DIR__ . "/imagemap.js");
        print "\n";
        print 'var module = ' . $this->getJavascriptModuleObjectName();
        print "\n";
        $this->activateMapScripts($keyLabelCodeMap);
        echo '</script>';
```

File: imagemapfunctions.js
Comment (steentje34) [inline imagemapfunctions.js]: Could you explain why this line is needed? I don't remember that this was required in the previous version, and I'd just like to understand why it is necessary.

**Review by steentje34 — COMMENTED** [review]
Hi @grezniczek, many thanks for your contribution! I reviewed the proposed changes, added some comments/questions, but I also have a general question:

Could I remove the .gitattributes, .gitignore & EM.code-workspace files after approving the PR? If I am not mistaken, these files contain some details specific to the configuration of your IDE? Correct me if I am mistaken, I am still a relative newbie with regards to GitHub.

Many thanks for your efforts!

File: imagemapfunctions.js
Comment (grezniczek) [reply imagemapfunctions.js]: > Could I remove the .gitattributes, .gitignore & EM.code-workspace files after approving the PR? If I am not mistaken, these files contain some details specific to the configuration of your IDE? Correct me if I am mistaken, I am still a relative newbie with regards to GitHub.

Yes, but they don't hurt, either, as the (content of the) `.gitattributes` file will ensure that releases won't include any of the files mentioned. Do as you like.

Comment (grezniczek) [reply imagemapfunctions.js]: > Could you explain why this line is needed? I don't remember that this was required in the previous version, and I'd just like to understand why it is necessary.

Since the JS code is inlined, I figured a good way to bring in the JavascriptModuleObject was through replacement. "JSMONAME" is just a placeholder. Your approach, i.e. bringing it in via assignment to `module`, works too, but in this case has the drawback of introducing a variable in the global scope. This should be avoided.

If you are referring to the `remove()` line ... the canvas will be lost at the proper place after MLM renders the page (it still exists, though), and thus any future clicks/updates will not register with the overlay library used. Thus, this forces the canvas to be recreated.

**Comment by grezniczek** [comment]
@steentje34 - With https://github.com/redcapuzgent/redcap_customimagemap/pull/4/commits/6739afcde335afd271aefd7b176bc5caf101e87a, I have now closed the version gap.

>Do you think this could also work? If yes, do you see advantages/disadvantages linked to both solutions?

I found no way around recreating the canvas (by removing it). I think the code is cleaner when the cases are handled in the JS file rather than create more complex JS in the PHP file.

**Comment by grezniczek** [comment]
I am still testing this. There may be some version gap remaining after all.

**Comment by grezniczek** [comment]
Ok, this should be done now. No version gap after all. I think this can be released now.

**Comment by steentje34** [comment]
@grezniczek, given your substantial contribution, would you want to be added as an author of the EM?

**Review by steentje34 — APPROVED** [review]

**Comment by grezniczek** [comment]
> @grezniczek, given your substantial contribution, would you want to be added as an author of the EM?

I don't mind either way. I'll leave it up to you.

**Pull request merged**
